File: backend/src/app/domain/models/db/lesson.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, JSON, String, Text, false, func
from sqlalchemy.orm import Mapped, mapped_column

from src.app.domain.models.db import Base
from src.app.domain.models.dto.lesson import LessonCaseDTO, LessonDTO, LessonSampleCaseDTO


class Lesson(Base):
    __tablename__ = "lessons"

    order: Mapped[str] = mapped_column(String(64))
    no_code: Mapped[bool] = mapped_column(Boolean(), default=False, server_default=false())
    slug: Mapped[str] = mapped_column(String(255), unique=True)
    name: Mapped[str] = mapped_column(String(255), default="Lesson name")
    body_markdown: Mapped[str] = mapped_column(Text(), default="body")
    code_editor_default: Mapped[str] = mapped_column(Text(), default="")
    cases: Mapped[list[dict[str, str | bool]]] = mapped_column(JSON(), default=list)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(),
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(),
        server_default=func.now(),
        onupdate=func.now(),
    )

    def to_dto(self) -> LessonDTO:
        """
        Convert model to DTO.

        :return: lesson dto
        """
        base_dto = LessonDTO.model_validate(obj=self, from_attributes=True)
        sample_cases = self._build_sample_cases(cases=base_dto.cases)

        return base_dto.model_copy(update={"sample_cases": sample_cases})

    @staticmethod
    def _build_sample_cases(cases: list[LessonCaseDTO]) -> list[LessonSampleCaseDTO]:
        """
        Build sample cases list from visible cases.

        :param cases: lesson cases

        :return: sample cases
        """
        samples = []
        for case in cases:
            if case.hidden:
                continue
            samples.append(
                LessonSampleCaseDTO(
                    name=case.name,
                    label=case.label,
                ),
            )

        return samples

File: frontend/src/shared/model/lesson.ts
export interface LessonApiResponse {
  id: string;
  slug: string;
  name: string;
  no_code: boolean;
  body_markdown: string;
  code_editor_default: string;
  cases: LessonCase[] | null;
  questions: LessonQuestion[] | null;
  sample_cases: LessonSampleCase[] | null;
  created_at: string;
  updated_at: string | null;
  order: string;
}

export interface LessonCase {
  name: string;
  label: string;
  script: string;
  hidden: boolean;
}

export interface LessonQuestion {
  prompt: string;
  options: string[];
  correct_option: number;
}

export interface LessonSampleCase {
  name: string;
  label: string;
}

export interface Lesson {
  id: string;
  slug: string;
  title: string;
  noCode: boolean;
  bodyMarkdown: string;
  codeEditorDefault: string;
  cases: LessonCase[];
  questions: LessonQuestion[];
  sampleCases: LessonSampleCase[];
  createdAt: Date;
  updatedAt: Date | null;
  order: string;
}

export interface LessonFormValues {
  order: string;
  slug: string;
  noCode?: boolean;
  title: string;
  bodyMarkdown: string;
  codeEditorDefault: string;
  cases: LessonCase[];
  questions?: LessonQuestion[];
}

export interface LessonPayload {
  order: string;
  slug: string;
  name: string;
  body_markdown: string;
  code_editor_default: string;
  cases: LessonCase[];
  questions: LessonQuestion[];
}

export const mapLesson = (response: LessonApiResponse): Lesson => {
  const cases = response.cases ?? [];
  return {
    id: response.id,
    slug: response.slug,
    title: response.name,
    noCode: cases.length === 0,
    bodyMarkdown: response.body_markdown,
    codeEditorDefault: response.code_editor_default,
    cases,
    questions: response.questions ?? [],
    sampleCases: response.sample_cases ?? [],
    createdAt: new Date(response.created_at),
    updatedAt: response.updated_at ? new Date(response.updated_at) : null,
    order: response.order,
  };
};

export const mapLessonPayload = (values: LessonFormValues): LessonPayload => {
  return {
    order: values.order,
    slug: values.slug,
    name: values.title,
    body_markdown: values.bodyMarkdown,
    code_editor_default: values.codeEditorDefault,
    cases: values.cases,
    questions: values.questions ?? [],
  };
};
